toolchain/check/context.cpp

```diff
@@ -306,37 +309,138 @@ auto Context::LookupUnqualifiedName(Parse::NodeId node_id,
+
+  CARBON_DIAGNOSTIC(ClassInvalidMemberAccess, Error,
+                    "Cannot access {0} member `{1}` of type `{2}`.",
+                    std::string_view, SemIR::NameId, SemIR::TypeId);
+  CARBON_DIAGNOSTIC(ClassMemberDefinition, Note,
+                    "The {0} member `{1}` is defined here.", std::string_view,
+                    SemIR::NameId);
+
+  auto parent_type_id = class_info.self_type_id;
+  std::string_view access_desc =
+      access_kind == SemIR::AccessKind::Private ? "private" : "protected";
+
+  if (access_kind == SemIR::AccessKind::Private && is_parent_access) {
+    if (auto base_decl = context.insts().TryGetAsIfValid<SemIR::BaseDecl>(
+            class_info.base_id)) {
+      parent_type_id = base_decl->base_type_id;
+    }
+    if (auto adapt_decl = context.insts().TryGetAsIfValid<SemIR::AdaptDecl>(
+            class_info.adapt_id)) {
+      parent_type_id = adapt_decl->adapted_type_id;
+    }
+
+    // Make sure that `parent_type_id` is replaced if this is a parent access.
+    CARBON_CHECK(parent_type_id != class_info.self_type_id);
+  }
+
+  context.emitter()
+      .Build(node_id, ClassInvalidMemberAccess, access_desc, name_id,
+             parent_type_id)
+      .Note(scope_result_id, ClassMemberDefinition, access_desc, name_id)
+      .Emit();
+}
+
+// Returns whether the access is prohibited by the access modifiers.
+static auto IsAccessProhibited(std::optional<AccessInfo> access_info,
+                               SemIR::AccessKind access_kind,
+                               bool is_parent_access) -> bool {
+  if (!access_info) {
+    return false;
+  }
+
+  switch (access_kind) {
+    case SemIR::AccessKind::Public:
+      return false;
+    case SemIR::AccessKind::Protected:
+      return access_info->highest_allowed_access == SemIR::AccessKind::Public;
+    case SemIR::AccessKind::Private:
+      return access_info->highest_allowed_access !=
+                 SemIR::AccessKind::Private ||
+             is_parent_access;
   }
-  return SemIR::InstId::Invalid;
 }
 
+// Information regarding a prohibited access.
+struct ProhibitedAccessInfo {
+  // The resulting inst of the lookup.
+  SemIR::InstId scope_result_id;
+  // The access kind of the lookup.
+  SemIR::AccessKind access_kind;
+  // If the lookup is from an extended scope. For example, if this is a base
+  // class member access from a class that extends it.
+  bool is_parent_access;
+};
+
 auto Context::LookupQualifiedName(Parse::NodeId node_id, SemIR::NameId name_id,
-                                  LookupScope scope, bool required)
+                                  LookupScope scope, bool required,
+                                  std::optional<AccessInfo> access_info)
     -> LookupResult {
   llvm::SmallVector<LookupScope> scopes = {scope};
+
+  // TODO: Support reporting of multiple prohibited access.
+  llvm::SmallVector<ProhibitedAccessInfo> prohibited_accesses = {};
```

Comment:
```suggestion
  llvm::SmallVector<ProhibitedAccessInfo> prohibited_accesses;
```

There's no need to explicitly initialize as empty.